### DISCUSSION TITLE
Nodejs example

### DISCUSSION THREAD

**Discussion by X7md** [termux/termux-gui#21] [Q&A]
I want nodejs example to how set UnixSocket and pass it to Termux:GUI...

**Comment by tareksander** [comment]
I tried to document the process in [Protocol.md](https://github.com/termux/termux-gui/blob/main/Protocol.md), specifically the "Connection" section. If you want to use Termux:GUI with Node, you should make a library for that, so you don't have to write special code for every project. You should already be familiar with Node and Javascript in general. I found these functions that can help you:  
[Writing a 32 bit big-endian into a Buffer](https://nodejs.org/api/buffer.html#bufwriteint32bevalue-offset)  
[Creating a UNIX socket](https://nodejs.org/api/net.html#serverlistenpath-backlog-callback)  
[Calling a program from Node](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback)

**Reply by X7md** [reply]
Still I don't get it, most example I have seen It bind to file path or ip address... I don't know how can I pass socket name 😓

**Reply by tareksander** [reply]
To use a socket name you should use a "path" that is a null byte ("\0") and then the socket name.

**Reply by X7md** [reply]
I tried but I got an error with event socket:

https://github.com/X7md/node-termuxGUI/blob/main/index.mjs

```js
$ am broadcast --user 0 -n com.termux.gui/.GUIReceiver --es mainSocket test --es eventSocket test2
now listening
node:events:504
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE: address already in use test2
    at Server.setupListenHandle [as _listen2] (node:net:1313:21)
    at listenInCluster (node:net:1378:12)
    at Server.listen (node:net:1476:5)
    at file:///data/data/com.termux/files/home/jsx-test/index.mjs:10:19
    at ModuleJob.run (node:internal/modules/esm/module_job:197:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:337:24)
    at async loadESM (node:internal/process/esm_loader:88:5)
    at async handleMainPromise (node:internal/modules/run_main:61:12)
Emitted 'error' event on Server instance at:
    at emitErrorNT (node:net:1357:8)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  code: 'EADDRINUSE',
  errno: -98,
  syscall: 'listen',
  address: '\x00test2',
  port: -1
}
```

**Reply by tareksander** [reply]
I recommend generating the socket names randomly to avoid collisions. Is this running in the same instance of node as your last test? If yes, you probably didn't close the socket, linux removes abstract sockets automatically when closed.
